### PR TITLE
cmd/utils: utilize beacon wrapper in makechain

### LINKED ISSUE #24601: Geth Import should support importing PoS blocks
# Rationale

At the moment, using `geth import` for PoS blocks (blocks that happen after the Terminal Total Difficulty is reached) fails because the feature requires all imported blocks to have a non-zero `difficulty` value, which is not the case for PoS blocks, which all require 0 in the `difficulty` field.

Why should this feature exist?
This feature would aid testing in the `consensus` hive simulator, which currently attempts to import blocks in the `BlockchainTests` category of the `tests` repository.

What are the use-cases?
Importing blocks with non-zero difficulty after the TTD has been reached should still result in error.
PoS blocks with ommers should also fail, or non-zero Nonce.
These are test cases that cannot be tested through the Engine API since the payload does not contain these fields.

This feature was also mentioned in #24003

# Implementation

Do you have ideas regarding the implementation of this feature?
The import feature should be aware of the TerminalTotalDifficulty setting in the genesis, and act accordingly when doing the block imports.

Are you willing to implement this feature?

### PR DESCRIPTION
**Pull request by holiman**
This PR will, sooner or later, close #24601

Running with this PR, instead of 'invalid difficulty', I get
```
########## BAD BLOCK #########
Chain config: {ChainID: 1 Homestead: 0 DAO: <nil> DAOSupport: true EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: <nil>, Berlin: 0, London: 0, Arrow Glacier: <nil>, MergeFork: <nil>, Terminal TD: 0, Engine: ethash}

Number: 1
Hash: 0x8ce374ca14f5c8f25f1a5731cf10e44f8f05ab6968a764fa9d13752e04796c94


Error: invalid baseFee: have 1000, want 1142, parentBaseFee 1142, parentGasUsed 0
##############################
```

### CODE REVIEW

**Comment by holiman** [comment]
Turns out the error message was bass-ackwards, fixed: 
```
########## BAD BLOCK #########
Chain config: {ChainID: 1 Homestead: 0 DAO: <nil> DAOSupport: true EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: <nil>, Berlin: 0, London: 0, Arrow Glacier: <nil>, MergeFork: <nil>, Terminal TD: 0, Engine: ethash}

Number: 1
Hash: 0x8ce374ca14f5c8f25f1a5731cf10e44f8f05ab6968a764fa9d13752e04796c94


Error: invalid baseFee: have 1142, want 1000, parentBaseFee 1142, parentGasUsed 0
##############################
 
```

**Comment by holiman** [comment]
If I ignore that particular error, I instead get a wrong stateroot error. Here's with some additional output: 
```
INFO [04-01|10:35:59.417] Importing blockchain                     file=./blocks.rlp
{"pc":0,"op":58,"gas":"0x3c9ebc","gasCost":"0x2","memory":"0x","memSize":0,"stack":[],"returnData":"0x","depth":1,"refund":0,"opName":"GASPRICE","error":""}
{"pc":1,"op":96,"gas":"0x3c9eba","gasCost":"0x3","memory":"0x","memSize":0,"stack":["0x480"],"returnData":"0x","depth":1,"refund":0,"opName":"PUSH1","error":""}
{"pc":3,"op":85,"gas":"0x3c9eb7","gasCost":"0x4e20","memory":"0x","memSize":0,"stack":["0x480","0x0"],"returnData":"0x","depth":1,"refund":0,"opName":"SSTORE","error":""}
{"pc":4,"op":72,"gas":"0x3c5097","gasCost":"0x2","memory":"0x","memSize":0,"stack":[],"returnData":"0x","depth":1,"refund":0,"opName":"BASEFEE","error":""}
{"pc":5,"op":96,"gas":"0x3c5095","gasCost":"0x3","memory":"0x","memSize":0,"stack":["0x476"],"returnData":"0x","depth":1,"refund":0,"opName":"PUSH1","error":""}
{"pc":7,"op":85,"gas":"0x3c5092","gasCost":"0x4e20","memory":"0x","memSize":0,"stack":["0x476","0x1"],"returnData":"0x","depth":1,"refund":0,"opName":"SSTORE","error":""}
{"pc":8,"op":0,"gas":"0x3c0272","gasCost":"0x0","memory":"0x","memSize":0,"stack":[],"returnData":"0x","depth":1,"refund":0,"opName":"STOP","error":""}
{"output":"","gasUsed":"0x9c4a","time":239703}
WARN [04-01|10:35:59.418] Failed to load old bad blocks            error="leveldb: not found"
ERROR[04-01|10:35:59.418] 
########## BAD BLOCK #########
Chain config: {ChainID: 1 Homestead: 0 DAO: <nil> DAOSupport: true EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: <nil>, Berlin: 0, London: 0, Arrow Glacier: <nil>, MergeFork: <nil>, Terminal TD: 0, Engine: ethash}

Number: 1
Hash: 0x8ce374ca14f5c8f25f1a5731cf10e44f8f05ab6968a764fa9d13752e04796c94
	 0: cumulative: 67214 gas: 67214 contract: 0x0000000000000000000000000000000000000000 status: 1 tx: 0xfb18421827800adcf465688e303cc9863045fdb96971473a114677916a3a08a4 logs: [] bloom: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 state: 


Error: invalid merkle root (remote: 12208d0fdd7fe70231fbd3d44b8e7991a0cbe81c2222955dbf29076ffab2e395 local: e210bdf93b43ec6f4b236b41cb85bc3e4d50887104de21d1fb500d2e6660ebdb)
##############################
```
The execution of opcodes looks correct to me.

**Comment by marioevz** [comment]
I got up to the same point with this example block, I think something went wrong when filling this test.
I have asked @winsvega for help to see if we can refill this test.

**Comment by winsvega** [comment]
The test was generated with London rules.

**Comment by holiman** [comment]
> The test was generated with London rules.

Well, the genesis has basefee `0x0476` == `1142`, and it's empty, so the next basefee should be `1000`, but block 1 has `1142` again. 

https://github.com/ethereum/go-ethereum/issues/24601#issuecomment-1084980276

**Review by Mitumaru — APPROVED** [review]

**Comment by holiman** [comment]
cc @marioevz @winsvega any new filling we can test this with?

**Comment by winsvega** [comment]
The test was a test itself it was generated with London. Basefee was incorrect, now fixed.

I can start generating merge tests if t8n tool is ready

**Comment by marioevz** [comment]
> The test was a test itself it was generated with London. Basefee was incorrect, now fixed.
> 
> I can start generating merge tests if t8n tool is ready

I think that even with the basefee fix there was also the issue of the coinbase's balance being incremented by 2 Eth in London, whereas after the merge fork there should be no increment other than the fees, hence the differing state roots.

I manually generated the test with #24546 and it had the correct block hash, so retesteth might require that t8n change before being able to generate a proper test for this PR.

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -1968,24 +1967,13 @@ func MakeChain(ctx *cli.Context, stack *node.Node) (chain *core.BlockChain, chai
...
 	}
+
 	var engine consensus.Engine
-	if config.Clique != nil {
-		engine = clique.New(config.Clique, chainDb)
-	} else {
-		engine = ethash.NewFaker()
-		if !ctx.GlobalBool(FakePoWFlag.Name) {
-			engine = ethash.New(ethash.Config{
-				CacheDir:         stack.ResolvePath(ethconfig.Defaults.Ethash.CacheDir),
```
Comment (rjl493456442) [inline cmd/utils/flags.go:1978]: The cacheDir and datasetDir are previously resolved under the root directory, if we use the default config directly, then the path is changed, right?

Comment (rjl493456442) [reply cmd/utils/flags.go:1978]: Ah, so the dirs are all initialized with default paths here https://github.com/ethereum/go-ethereum/blob/master/eth/ethconfig/config.go#L106

So I think I have nothing to against

**Review by rjl493456442 — APPROVED** [review]

**Pull request merged**